rpc/core/dev.go

```diff
@@ -1,12 +1,104 @@
 package core
 
 import (
+	"fmt"
+
+	abcitypes "github.com/tendermint/tendermint/abci/types"
 	ctypes "github.com/tendermint/tendermint/rpc/core/types"
 	rpctypes "github.com/tendermint/tendermint/rpc/jsonrpc/types"
+	"github.com/tendermint/tendermint/state/indexer"
+	"github.com/tendermint/tendermint/types"
 )
 
 // UnsafeFlushMempool removes all transactions from the mempool.
 func (env *Environment) UnsafeFlushMempool(ctx *rpctypes.Context) (*ctypes.ResultUnsafeFlushMempool, error) {
 	env.Mempool.Flush()
 	return &ctypes.ResultUnsafeFlushMempool{}, nil
 }
+
+// UnsafeReIndex re-index the block/transaction events into the eventsinks.
+func (env *Environment) UnsafeReIndex(
```

Comment:
But the indexer doesn't know anything about the blocks and states.

Yes, the re-index call will interleave with the regular indexer operations at the same time. Inside the Index functions, the kv/psql DB process the operations atomically. And there is no sharing instance (except the DB) in the Index functions so I think it would be okay?